components/eam/src/physics/cam/micro_p3.F90

```diff
@@ -2836,33 +2838,51 @@ subroutine back_to_cell_average(cld_frac_l,cld_frac_r,cld_frac_i,
 
 end subroutine back_to_cell_average
 
-subroutine prevent_ice_overdepletion(pres,t_atm,qv,latent_heat_sublim,inv_dt,    &
-   qidep,qi2qv_sublim_tend)
+subroutine prevent_ice_overdepletion(pres,t_atm,qv,latent_heat_vapor,latent_heat_sublim,inv_dt,dt,qidep,qinuc,    &
+   qr2qv_evap_tend,qi2qv_sublim_tend)
 
-   !-- Limit ice process rates to prevent overdepletion of sources such that
-   !   the subsequent adjustments are done with maximum possible rates for the
-   !   time step.  (note: most ice rates are adjusted here since they must be done
-   !   simultaneously (outside of iice-loops) to distribute reduction proportionally
-   !   amongst categories.
-   !PMC - might need to rethink above statement since only one category now.
+   !-- Limit evaporative processes to prevent the QV field from ice supersaturation (evaporation should not lead to supersaturation) 
 
    implicit none
 
    real(rtype), intent(in) :: pres
    real(rtype), intent(in) :: t_atm
    real(rtype), intent(in) :: qv
-   real(rtype), intent(in) :: latent_heat_sublim
-   real(rtype), intent(in) :: inv_dt
-
-   real(rtype), intent(inout) :: qidep
-   real(rtype), intent(inout) :: qi2qv_sublim_tend
-
-   real(rtype) :: dumqv_sat_i, qdep_satadj
-
-   dumqv_sat_i = qv_sat(t_atm,pres,1)
-   qdep_satadj = (qv-dumqv_sat_i)/(1._rtype + bfb_square(latent_heat_sublim)*dumqv_sat_i/(cp*rv* bfb_square(t_atm) ))*inv_dt
-   qidep  = qidep*min(1._rtype,max(0._rtype, qdep_satadj)/max(qidep, 1.e-20_rtype))
-   qi2qv_sublim_tend  = qi2qv_sublim_tend*min(1._rtype,max(0._rtype,-qdep_satadj)/max(qi2qv_sublim_tend, 1.e-20_rtype))
+   real(rtype), intent(in) :: latent_heat_sublim,latent_heat_vapor
+   real(rtype), intent(in) :: inv_dt,dt
+   real(rtype), intent(in) :: qidep,qinuc
+
+   real(rtype), intent(inout) :: qi2qv_sublim_tend,qr2qv_evap_tend
+
+   real(rtype) :: dumqv_sat_i, qdep_satadj, qrevp_satadj
+   real(rtype) :: qtmp_all,ttmp_all,qv_sat_l,q_sink,t_sink,dumqv_sat_l,qv_source_evp,qi2qv_sublim_satadj
+
+   ! ... available water vapor 
+   qtmp_all = qv - (qidep + qinuc)*dt + (qi2qv_sublim_tend + qr2qv_evap_tend)*dt 
+   ! ... corresponding temperature
+   ttmp_all = t_atm + ((qidep-qi2qv_sublim_tend+qinuc)*latent_heat_sublim*inv_cp + (-qr2qv_evap_tend*latent_heat_vapor*inv_cp))*dt
+
+   ! ... water vapor needed for liquid saturation is the minimal requirment for both liquid and ice sturation
+   qv_sat_l = qv_sat(ttmp_all,pres,0)
+
+   ! ... if water vapor mass exceeds liquid saturation, we limit only source terms (e.g., sublimation, rain evp) 
+   if(qtmp_all > qv_sat_l)then
```

Comment:
if statements don't work for GPU calculations. The original max and min approach is probably needed to ensure BFB agreement between Kokkos and F90 implementations.